fast64_internal/oot/oot_upgrade.py

```diff
@@ -35,3 +47,167 @@ def upgradeRoomHeaders(roomObj: Object, objData: OoT_ObjectData):
+                    value = int(getEvalParams(data[oldName]), base=16)
+
+                data[newName] = value
+
+            del data[oldName]
+
+
+def convertOldDataToEnumData(data, oldDataToEnumData: list[tuple[str, str, list[tuple[str, str, str]]]]):
+    # conversion to another prop type
+    for (oldName, newName, enumList) in oldDataToEnumData:
+        if oldName in data:
+            # get the old data
+            oldData = data[oldName]
+
+            # if anything goes wrong there set the value to custom to avoid any data loss
+            try:
+                if isinstance(oldData, str):
+                    # get the value, doing an eval for strings
+                    # account for custom elements in the enums by adding 1
+                    value = int(getEvalParams(oldData), base=16) + 1
+
+                    # special cases for ocarina action enum
+                    # since we don't have everything the value need to be shifted
+                    if newName == "ocarinaAction":
+                        if value in [0x00, 0x01, 0x0E] or value > 0x1A:
+                            raise IndexError
+
+                        if value > 0x0E:
+                            value -= 1
+
+                        value -= 2
+
+                    if newName == "csSeqID":
+                        # the old fade out value is wrong, it assumes it's a seq id
+                        # but it's not, it's a seq player id,
+                        # hence why we raise an error so it defaults to "custom" to avoid any data loss
+                        # @TODO: find a way to check properly which seq command it is
+                        raise NotImplementedError
+                elif isinstance(oldData, int):
+                    # account for custom elements in the enums by adding 1
+                    value = oldData + 1
+
+                    # another special case, this time for the misc enum
+                    if newName == "csMiscType":
+                        if value in [0x00, 0x04, 0x05]:
+                            raise IndexError
+
+                        if value > 0x05:
+                            value -= 2
+
+                        value -= 1
+                else:
+                    raise NotImplementedError
+
+                # if the value is in the list find the identifier
+                if value < len(enumList):
+                    setattr(data, newName, enumList[value][0])
+                else:
+                    # else raise an error to default to custom
+                    raise IndexError
+            except:
+                setattr(data, newName, "Custom")
+                setattr(data, f"{newName}Custom", str(oldData))
+
+                # @TODO: find a way to check properly which seq command it is
+                if newName == "csSeqID":
+                    setattr(data, "csSeqPlayer", "Custom")
+                    setattr(data, "csSeqPlayerCustom", str(oldData))
```

Comment:
can't say I understand the todo

like, the seqId value here may be `seqPlayer` from `CS_FADE_OUT_SEQ` ?

Reply:
I should rephrase that but we need a proper way to check if we're dealing with the fade out command or the start/stop one, there's a minor difference between these two, the fade out one doesn't take a seq id but the player type (as defined in sequence.h iirc)

all of this is based on the informations inside z64cutscene_commands.h btw

the main issue at the moment is that we have to use "Custom" when upgrading because everything was assumed to have the same parameters, and now we know for sure this is a wrong assumption